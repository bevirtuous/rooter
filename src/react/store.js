import history from 'history/browser';
import create from 'zustand';

function createStoreEntry(location) {
  const { hash, key, pathname, search, state } = location;
  return {
    fragment: hash,
    id: key || 'default',
    pathname,
    search,
    state: state || {},
  };
}

export default create((set) => {
  const route = createStoreEntry(history.location);

  return {
    current: 0,
    routes: [route],
    back: () => set((state) => ({
      current: Math.max(0, state.current - 1),
    })),
    forward: () => set((state) => ({
      current: Math.min(state.routes.length - 1, state.current + 1),
    })),
    push: (location) => set((state) => {
      const nextCurrent = state.current + 1;
      const nextRoute = state.routes[nextCurrent];

      if (nextRoute && nextRoute.id === location.key) {
        return state.forward(location);
      }

      const nextRoutes = state.routes
        .slice(0, nextCurrent)
        .concat(createStoreEntry(location));

      return {
        current: nextCurrent,
        routes: nextRoutes,
      };
    }),
  };
});
