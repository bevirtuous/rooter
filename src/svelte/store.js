import { writable } from 'svelte/store';
import { emitter, EVENT, UPDATE } from '../core';

const store = writable({});

emitter.on(EVENT, (payload) => {
  if (payload.action === UPDATE) {
    store.set(payload.route);
  } else {
    store.set(payload.next);
  }
});

export default store;
