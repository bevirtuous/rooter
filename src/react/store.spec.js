import store from './store';

jest.mock('history/browser', () => ({
  location: {
    pathname: '/hello/world',
    search: '?param=123',
    hash: 'title',
    state: null,
    key: 'default'
  }
}))

describe('React - Store', () => {
  beforeAll(() => {
    const { push } = store.getState();
    push({
      pathname: '/second/route',
      search: '',
      hash: '',
      key: '#2'
    });
  });

  it('should correctly create the store with initial history entry', () => {
    expect(store.getState()).toMatchSnapshot();
  });

  it('should allow back', () => {
    const { back } = store.getState();
    store.setState({ current: 2 });
    back();
    expect(store.getState().current).toBe(1);
  });

  it('should not go back past limit', () => {
    const { back } = store.getState();
    store.setState({ current: 0 });
    back();
    expect(store.getState().current).toBe(0);
  });

  it('should allow back', () => {
    const { forward } = store.getState();
    store.setState({ current: 0 });
    forward();
    expect(store.getState().current).toBe(1);
  });
  
  it('should not go forward past limit', () => {
    const { forward } = store.getState();
    store.setState({ current: 1 });
    forward();
    expect(store.getState().current).toBe(1);
  });

  it('should allow push', () => {
    const { push } = store.getState();
    store.setState({ current: 1 });
    push({
      pathname: '/hello/world/again',
      search: '',
      hash: '',
      state: {
        hi: '123',
      },
      key: '#3'
    });
    expect(store.getState()).toMatchSnapshot();
  });

  it('should allow push and remove forward routes', () => {
    const { push } = store.getState();
    store.setState({ current: 0 });
    push({
      pathname: '/hello/world/again',
      search: '',
      hash: '',
      state: {
        hi: '123',
      },
      key: '#3'
    });
    expect(store.getState()).toMatchSnapshot();
  });

  it('should forward instead of push', () => {
    const { push } = store.getState();
    store.setState({ current: 0 });

    push({
      pathname: '/hello/world/again',
      search: '',
      hash: '',
      key: '#3'
    });

    const { current, routes } = store.getState();
    expect(routes[current].state).toEqual({ hi: '123' });
  });
});
