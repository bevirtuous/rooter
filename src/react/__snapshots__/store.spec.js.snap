// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React - Store should allow push 1`] = `
Object {
  "back": [Function],
  "current": 2,
  "forward": [Function],
  "push": [Function],
  "routes": Array [
    Object {
      "fragment": "title",
      "id": "default",
      "pathname": "/hello/world",
      "search": "?param=123",
      "state": Object {},
    },
    Object {
      "fragment": "",
      "id": "#2",
      "pathname": "/second/route",
      "search": "",
      "state": Object {},
    },
    Object {
      "fragment": "",
      "id": "#3",
      "pathname": "/hello/world/again",
      "search": "",
      "state": Object {
        "hi": "123",
      },
    },
  ],
}
`;

exports[`React - Store should allow push and remove forward routes 1`] = `
Object {
  "back": [Function],
  "current": 1,
  "forward": [Function],
  "push": [Function],
  "routes": Array [
    Object {
      "fragment": "title",
      "id": "default",
      "pathname": "/hello/world",
      "search": "?param=123",
      "state": Object {},
    },
    Object {
      "fragment": "",
      "id": "#3",
      "pathname": "/hello/world/again",
      "search": "",
      "state": Object {
        "hi": "123",
      },
    },
  ],
}
`;

exports[`React - Store should correctly create the store with initial history entry 1`] = `
Object {
  "back": [Function],
  "current": 1,
  "forward": [Function],
  "push": [Function],
  "routes": Array [
    Object {
      "fragment": "title",
      "id": "default",
      "pathname": "/hello/world",
      "search": "?param=123",
      "state": Object {},
    },
    Object {
      "fragment": "",
      "id": "#2",
      "pathname": "/second/route",
      "search": "",
      "state": Object {},
    },
  ],
}
`;
